### PR TITLE
Snow 1065172 gh workflow optimization

### PR DESCRIPTION
**Pull request by sfc-gh-mraba**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-1065172

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [x] I am solving issues with long running and failing jobs

3. Please describe how your code solves the related issue.

- we have refactored connection creation to be created with fixtures.
- we have split engine creation from url creation
- we have moved from tox to hatch
- we have split tests to be run in separate jobs to speed up checks
- we have dropped unused codecov

### CODE REVIEW

File: .github/workflows/build_test.yml [hunk 1]
```diff
@@ -24,122 +24,164 @@ jobs:
...
+        run: |
+          gpg --quiet --batch --yes --decrypt --passphrase="$PARAMETERS_SECRET" \
+          .github/workflows/parameters/parameters_${{ matrix.cloud-provider }}.py.gpg > tests/parameters.py
+      - name: Run tests
+        run: hatch run test-dialect
+      - uses: actions/upload-artifact@v4
+        with:
+          name: coverage.xml_dialect-${{ matrix.os }}-${{ matrix.python-version }}-${{ matrix.cloud-provider }}
+          path: |
+            ./coverage.xml
```
Comment (sfc-gh-astus) [inline .github/workflows/build_test.yml:93]: upload is not needed if we don't have coverage

File: .github/workflows/build_test.yml [hunk 2]
```diff
@@ -24,122 +24,164 @@ jobs:
...
 
-  combine-coverage:
-    if: ${{ success() || failure() }}
-    name: Combine coverage
-    needs: [test]
-    runs-on: ubuntu-latest
+  test-dialect-compatibility:
+    name: Test dialect compatibility ${{ matrix.os }}-${{ matrix.python-version }}-${{ matrix.cloud-provider }}
+    # needs: test-dialect
```
Comment (sfc-gh-astus) [inline .github/workflows/build_test.yml:97]: Can be removed

File: .github/workflows/build_test.yml [hunk 3]
```diff
@@ -24,122 +24,164 @@ jobs:
...
-      - name: Upgrade setuptools, pip and wheel
-        run: python -m pip install -U setuptools pip wheel
-      - name: Install tox
-        run: python -m pip install tox
+      - name: Upgrade and install tools
+        run: |
+          python -m pip install -U pip
+          python -m pip install -U hatch
+          python -m hatch env create default
+          python -m hatch env find
```
Comment (sfc-gh-astus) [inline .github/workflows/build_test.yml:39]: Is this needed?

File: .github/workflows/build_test.yml [hunk 4]
```diff
@@ -24,122 +24,164 @@ jobs:
...
-          path: .tox/htmlcov
-      - name: Publish xml coverage
-        uses: actions/upload-artifact@v2
+          name: coverage.xml_dialect-compatibility-${{ matrix.os }}-${{ matrix.python-version }}-${{ matrix.cloud-provider }}
+          path: |
+            ./coverage.xml
+
+  test-dialect-run-v20:
+    name: Test dialect run v20 ${{ matrix.os }}-${{ matrix.python-version }}-${{ matrix.cloud-provider }}
+    # needs: test-dialect
```
Comment (sfc-gh-astus) [inline .github/workflows/build_test.yml:144]: Can be removed

File: tests/conftest.py [hunk 5]
```diff
@@ -102,23 +105,20 @@ def help():
 
 @pytest.fixture(scope="session")
 def db_parameters():
-    return get_db_parameters()
+    yield get_db_parameters()
 
 
-def get_db_parameters():
+def get_db_parameters() -> dict:
```
Comment (sfc-gh-turbaszek) [inline tests/conftest.py:111]:
```suggestion
def get_db_parameters() -> Dict:
```

File: .github/workflows/build_test.yml (hunk 1 above)
Comment (sfc-gh-mraba) [reply .github/workflows/build_test.yml:93]: You're right.
I wonder id we should have raport to look at for future PR's

File: .github/workflows/build_test.yml (hunk 3 above)
Comment (sfc-gh-mraba) [reply .github/workflows/build_test.yml:39]: The find option can be removed.
I'd like to separate env creation from tests to make output clean.

**Review by sfc-gh-astus — APPROVED** [review]

**Pull request merged**